Improve partial message support in logger

anusha-ragunathan opened the pull request:
Improve partial message support in logger and use it in Splunk.

Docker daemon has a 16K buffer for log messages. If a message length
exceeds 16K, it should be split by the logger and merged at the
endpoint.

This change adds `LogMetaData` struct for enhanced partial support
IsPartial (bool)    : Indicates if this message is partial
LastPartial (bool) : indicates if this is the last of all partials.
ID (string)            : unique 32 bit ID. ID is same across all partials.
Ordinal (int starts at 1) : indicates the position of msg in the series of partials.
 
  Signed-off-by: Anusha Ragunathan <anusha.ragunathan@docker.com>

anusha-ragunathan committed:
Improve partial message support in logger

Docker daemon has a 16K buffer for log messages. If a message length
exceeds 16K, it should be split by the logger and merged at the
endpoint.

This change adds `PartialLogMetaData` struct for enhanced partial support
- LastPartial (bool) : indicates if this is the last of all partials.
- ID (string)        : unique 32 bit ID. ID is same across all partials.
- Ordinal (int starts at 1) : indicates the position of msg in the series of partials.
Also, the timestamps across partials in the same.

Signed-off-by: Anusha Ragunathan <anusha.ragunathan@docker.com>

cpuguy83 commented:
> daemon/logger/splunk/splunk.go:123::error: struct field tag `json: "partial,omitempty"` not compatible with reflect.StructTag.Get: bad syntax for struct tag value (vet)

thaJeztah commented on daemon/logger/splunk/splunk.go:
```diff
@@ -120,6 +120,7 @@ type splunkMessage struct {
 	Source     string      `json:"source,omitempty"`
 	SourceType string      `json:"sourcetype,omitempty"`
 	Index      string      `json:"index,omitempty"`
+	Partial    bool        `json: "partial,omitempty"`
```
Ah, yes the space after `json:` should be removed

anusha-ragunathan commented:
powerpc and z tests failures due to https://github.com/moby/moby/issues/35775

cpuguy83 reviewed (approved):
LGTM

cpuguy83 commented:
@thaJeztah Who do we know from plunk to review this?

thaJeztah commented:
@outcoldman and @glennblock worked on the logging driver previously, but don't think they're still working for Splunk (but perhaps they know who to contact), and I saw @michaelwilde comment on one of the pull requests (he looks to be still with Splunk, so 👍)

thaJeztah commented:
oh! saw https://github.com/splunk/docker-logging-plugin, so YOLO let me spam @bbourbie and @hyfather 😇

outcoldman commented:
Also @mchene can find somebody for review.

Btw, I believe HTTP Event Collector will just ignore Partial field in the event, see REST endpoint definition http://docs.splunk.com/Documentation/Splunk/7.0.1/RESTREF/RESTinput#services.2Fcollector

hyfather commented:
Sorry just looking at this, was OOO -- hope everyone had a great holiday! 🎉 

Each HTTP request to Splunk is treated separately, so there needs to be a unique identifier across all partial requests that docker sends so that Splunk can tell which lines need to be merged into a single event. Note that the Splunk endpoint that docker is making requests to could also be a load balancer and multiple HTTP requests carrying a partial line could be send to separate Splunk indexing servers.

Getting a little Splunk specific here -- even if we change the plugin to add a line merging token to the first of the partial request to use the `SHOULD_LINEMERGE` & `BREAK_ONLY_BEFORE` directives there is no guarantee that two partial requests will hit the same indexer. Any ideas on how to solve for this using `props.conf`?

As far as sending `partial` with each request, looks like HEC doesn't like it for a default `JSON` style payload --

```
$ curl -d '{"event": {"foo": "bar"}}' http://localhost:8088/services/collector?channel=20337e7a-eb6e-46f4-94a8-926b90e11bc5 -H 'Authorization: Splunk 20337e7a-eb6e-46f4-94a8-926b90e11bc5'
=> {"text":"Success","code":0,"ackId":0}


$ curl -d '{"partial": false, "event": {"foo": "bar"}}' http://localhost:8088/services/collector?channel=20337e7a-eb6e-46f4-94a8-926b90e11bc5 -H 'Authorization: Splunk 20337e7a-eb6e-46f4-94a8-926b90e11bc5'
=> {"text":"No data","code":5}
```

outcoldman commented:
@hyfather I am confused, looking on the documentation http://docs.splunk.com/Documentation/Splunk/7.0.1/RESTREF/RESTinput#services.2Fcollector, does not seem like HTTP Event Collector supports `partial` field. If this is a new, additional feature, is it going to be documented? Also if this is a new feature in 7.x, I assume that will not work with Splunk versions below 7?

hyfather commented:
@outcoldman `partial` isn't a field on the HEC. It is this PR that adds `partial` to the outbound HTTP request. Maybe I got the structure of how it'll be marshaled wrong?

cpuguy83 commented:
Maybe some missing context here, but I believe Docker Inc has been working with Splunk to get some kind of supported added for this scenario (single line broken into multiple events due to huge lines) so that the moby splunk driver can deal with it properly.

hyfather commented:
@cpuguy83 got it! I wasn't aware of this -- it's probably best to wait for @mchene to weigh in. Thanks!

boaz0 reviewed (approved):
LGTM 👍

anusha-ragunathan commented:
Looks like without unique ID attached to messages, this PR would not be useful for Splunk. So lets not merge this yet.

@hyfather: Adding a unique 32 bit string ID to the messages can be done alongside partial log detection. Diff on top of this PR: https://gist.github.com/anusha-ragunathan/214b5c93ee89cb8342d101ada0f8bc63
Would that work?

anusha-ragunathan commented:
More failures on ppc
https://jenkins.dockerproject.org/job/Docker-PRs-powerpc/7746/consoleText

```
=== RUN   TestDaemonRestartKillContainers/live-restore=true/container_with_restart=always/stop-daemon
--- FAIL: TestDaemonRestartKillContainers (0.00s)
    --- PASS: TestDaemonRestartKillContainers/live-restore=true/container_with_restart=always/kill-daemon (2.20s)
    	daemon.go:285: [dc7f9929f10b5] waiting for daemon to start
    	daemon.go:317: [dc7f9929f10b5] daemon started
    	daemon.go:285: [dc7f9929f10b5] waiting for daemon to start
    	daemon.go:317: [dc7f9929f10b5] daemon started
    	daemon.go:275: [dc7f9929f10b5] exiting daemon
    --- FAIL: TestDaemonRestartKillContainers/live-restore=false/container_with_restart=always/kill-daemon (2.37s)
    	daemon.go:285: [df3331f4626eb] waiting for daemon to start
    	daemon.go:285: [df3331f4626eb] waiting for daemon to start
    	daemon.go:285: [df3331f4626eb] waiting for daemon to start
    	daemon.go:275: [df3331f4626eb] exiting daemon
    	daemon.go:203: Error starting daemon with arguments: []
    --- FAIL: TestDaemonRestartKillContainers/live-restore=false/container_with_restart=always/stop-daemon (0.31s)
    	daemon.go:285: [d66d385ff1cb0] waiting for daemon to start
    	daemon.go:275: [d66d385ff1cb0] exiting daemon
    	daemon.go:203: Error starting daemon with arguments: []
```

anusha-ragunathan commented:
@cpuguy83 @thaJeztah @outcoldman @hyfather @ripcurld0  I've updated the PR to include a unique `PartialID` for partial messages so that they can be assembled by Splunk. PTAL.

cpuguy83 commented:
I think for this to work we'd need a message ID which span across multiple partials and then an ordinal for each partial.

anusha-ragunathan commented:
My bad. Right now, `msg.reset` will reset the PartialID on each partial `Log`. Let me fix that and add a sequence number to the partial messages.
The problem of partials getting directed to different indexing servers still remain.

hyfather commented:
@cpuguy83, @anusha-ragunathan -- yes a `PartialID` with an ordinal makes the most sense, along with the first partial indicating the total pieces to expect (is this possible?). This should be sufficient information for any aggregation service to assemble events or discard incomplete ones. 
In addition, I'll follow up internally @ Splunk to see who is working on the receiving piece.

cpuguy83 commented:
I don't think we can send along the number to expect, which indeed would be nice, unless we cache all the messages locally first, this would in turn bring us back to square one with the size of these log messages.

hyfather commented:
Ah got it. Would it make sense to send a special identifier with the last part so that the joined event can be finalized on the receiving side? I suppose even an empty message with an incremented ordinal number would work for this.

cpuguy83 commented:
Yes, an "end of message" flag would be helpful.

cpuguy83 commented:
We should get some other input on this from the community.

anusha-ragunathan commented:
@cpuguy83 @hyfather : Added ordinal and last partial detection. PTAL

anusha-ragunathan commented:
Adding  some members from the log driver community for input on "Improved Partial Log support" proposed in this PR.

cc @jahkeup @samuelkarp from AWS Logs.

thaJeztah commented:
Let me reel in some more people (yay!) that worked on other logging drivers;

@mikedanese @farcaller @gjaskiewicz (GCP Logging)

@tagomoris @kzk (Fluentd)

@mariussturm (Gelf logging driver)

We're looking for input on the way we'll be handling partial messages; hope you are able to help with that, to verify if the design proposed here would work (also for future addition to other logging drivers). Perhaps there are already standards/approaches that we could apply?

anusha-ragunathan commented:
Ping?

hyfather commented:
@mchene is the point of contact from Splunk.

samuelkarp commented:
I don't believe this is particularly useful for the awslogs driver unless there's a plan to start sending out-of-order messages to the driver (I hope not!).  CloudWatch Logs already has a few maximums that are dealt with by the log driver, and https://github.com/moby/moby/pull/34888 fixed the logger package's buffer to use the same size as the driver so that we don't get any partials that need to be reassembled anyway.

anusha-ragunathan commented:
@cpuguy83 @thaJeztah : PTAL

cpuguy83 reviewed (commented):
Design seems ok.
Only thing is I'd probably replace the `partial` bool in the `LogMessage` type with a struct containing the details of the partial.

Is splunk supporting this?
I'd like to make sure we aren't send out data to splunk until it's fully supported (and the right data model).

cpuguy83 commented:
@samuelkarp

> unless there's a plan to start sending out-of-order messages to the driver (I hope not!)

I can't see any reason we'd want to do this. I think the implementation is just to support backends that could end up with out of order messages due to the log service's implementation.

anusha-ragunathan commented:
@hyfather : Unless we get their engineers to review and agree that Splunk is okay with the design and will implement the corresponding data model,  this PR cannot be merged. Can you poke the appropriate guys on your side?

hyfather commented:
@anusha-ragunathan -- will do, thanks for holding this off.

chenziliang commented:
It is very hard for Splunk Enterprise / Cloud to do the event assembly in a distributed env. Docker driver / plugin shall do the assembly and present an intact event to Splunk HEC

anusha-ragunathan commented:
@chenziliang : It doesnt make sense for the docker splunk driver to reassemble it. The reason is, it is compiled into the docker daemon and reassembling the chunks beats the purpose of splitting the message in the first place.

It does make sense to do this in the splunk logging plugin.

chenziliang commented:
Before sending events to Splunk HEC, can Spink docker driver do the assembling ?

chenziliang commented:
May I ask where the split is happens ?

chenziliang commented:
For Splunk docker driver, I intend to think splitting the event probably can not help from lowering the memory consumption as it is best to do some event batch in Splunk docker driver to make the logging driver performant which in turn low the memory footprint ?

If the Splunk docker driver does event post one by one, the performance will be far worse which results in a far slower consumer which I don’t know if it will cause dockerd(the producer) hang or use more memory?

cpuguy83 commented:
@chenziliang Sure, but if an event has an unbounded size then any container can crash the system regardless of memory restrictions on the container.

chenziliang commented:
Would like to know why supporting unlimited size of log ?

cpuguy83 commented:
@chenziliang We don't, and that's the problem here.

chenziliang commented:
Wondering if dockerd can provide a max size log flag to limit the max size of a single event and by default it is 1MB for instance which can probably serves 99.9 % use cases.

It doesn’t make too much sense to me to have a huge event since it is probably very hard to analyze and secondly Splunk will truncate single event which is large than 1MB with by default settings I think.

An unlimited log size looks a abnormal application behavior which should be best prevented by dockerd.

Just a few thoughts :).

cpuguy83 commented:
@chenziliang 

I agree this is due to poor application design and incorrect logging usage.... however it is a fact of life that we have had to deal with.

Docker does set a max size (16KB by default, I believe). When the size is met, the event is broken up into however many messages are required (docker looks for a new line character).
The issue is having it broken up is problematic for some people.

If splunk is truncating at 1MB anyway, maybe we can just set a custom buffer size for the splunk logger to 1MB.

chenziliang commented:
@cpuguy83 the key point here is "event reassembly in Splunk in a distributed env is really hard and probably is un-realistic" in short term. There are 3 ways to mitigate this problem:

1. dockerd limits the max size of a single event and doesn't do event break up. When event size is exceeding a max limit, dockerd truncates it.
2. dockerd limits the max size of a single event and does event break up. Splunk docker driver / plugin does the event assembling, Splunk docker driver does the truncation when a max size is hit (we don't want to post a 100MB single event to Splunk for example) to avoid memory outage.
3. dockerd limits the max size of a single event and does event break up. Splunk docker driver / plugin just bake the "split up" meta information into "fields" of Splunk JSON event. Splunk will not do the event assembly but with the "split up" information indexed into Splunk, user still have a chance to search all pieces of events.
More info for "fields" of Splunk JSON event 
http://dev.splunk.com/view/event-collector/SP-CAAAE6P#meta

chenziliang commented:
And increasing the max limit of 16KB to 1MB for a single event in dockerd should cover 99.9% use cases which don't need require event split.

anusha-ragunathan commented:
@thaJeztah @cpuguy83 : Ping

I've removed Splunk specifics from this PR. Splunk will not be using this change. Instead, they are on their way to implement a logging plugin, which gives them the flexibility of having a buffer size of choice.

anusha-ragunathan commented:
Fluentd will use the new metadata as part of 36159

anusha-ragunathan commented:
Janky failure due to https://github.com/moby/moby/issues/36611

anusha-ragunathan commented:
z failure due to https://github.com/moby/moby/issues/36547

thaJeztah commented on daemon/logger/copier.go:
```diff
@@ -112,7 +136,20 @@ func (c *Copier) copySrc(name string, src io.Reader) {
+						ordinal = 0
+						firstPartial = false
+					}
+					ordinal++
```
Is it correct that we increment here? So a full log-entry will have `MetaData.Ordinal: 0`, and the first part of a split message will have `MetaData.Ordinal: 1`?

anusha-ragunathan replied:
Yes. A full log entry will have a Ordinal 0 and an empty ID.

thaJeztah commented on daemon/logger/copier.go:
```diff
@@ -112,7 +136,20 @@ func (c *Copier) copySrc(name string, src io.Reader) {
+						firstPartial = false
+					}
+					ordinal++
+					setLogMetaData(msg, partialid, ordinal, false)
```
Looking at the code, I'm a bit in doubt if we should keep the `setLogMetaData()` function, or if it's clearer to just do it inline.

Was even looking if it's possible somehow to do `msg := NewMessage()` if we create a new log-entry, and keep the same `msg`, until we reached the end. This would allow

- Generating a unique ID inside `NewMessage()` - both partial and non-partial log-entries would get an ID
- Inside the loop do `msg.MetaData.ordinal++`
- no need keep track  / "reset" these local variables (`msg = NewMessage()` would reset them)

Haven't fully gone through that idea though to see if that's easy to refactor

anusha-ragunathan replied:
The design is such that IDs are handed out only to partials. Full log entries will have empty IDs.

thaJeztah replied:
Right, so wondering if it would be easier to process if an ID was always present (basically, a "full" entry is just a partial consisting of one part)

anusha-ragunathan replied:
Not necessary. An empty ID string is an indicator for a full log entry. Another way to do this would be to have a pointer to `PartialLogMetaData` and check for `nil` pointer to indicate a full log entry.

thaJeztah replied:
Let me try explaining my train of thought: if each message has a unique ID (irregardless of being partial or "full"), an aggregator could find all parts of a message by "grouping" by ID, and stitch them together based on `ordinal`.

The same logic could be used both for _partial_ and _full_ messages, so processing could be easier to handle if there's less differences between those

thaJeztah reviewed (changes requested):
left some thoughts, but let me know what you think

stevvooe commented on api/types/backend/backend.go:
```diff
@@ -26,6 +26,16 @@ type ContainerAttachConfig struct {
+// are split into chunks with this metadata. The expectation is for the logger endpoints to assemble
+// the chunks using this metadata.
+type LogMetaData struct {
+	IsPartial   bool
```
Can we document these fields?

```
ID string // identifies the log message or group of messages comprising a single record 
Ordinal // ordering of message in partial group
IsPartial bool // true if this message is made up of 2 or more messages 
LastPartial bool // true if this message is last of a partial
```

Also, we might be able to get rid of the need for the partial bool if we use 0 for complete message and 1 for first ordinal of partial, then all you need is `last` bool.

anusha-ragunathan replied:
Doc-ed and removed `FirstPartial`

stevvooe commented on daemon/logger/copier.go:
```diff
@@ -100,6 +113,17 @@ func (c *Copier) copySrc(name string, src io.Reader) {
+						ordinal++
+						setLogMetaData(msg, partialid, ordinal, true)
+
+						// reset
```
Are we supposed to reset after adding to the partial?

anusha-ragunathan replied:
This is the last chunk that's added. The reset is for the local variables used to track `first`, `hasMore` and `ordinal`

anusha-ragunathan commented on api/types/backend/backend.go:
```diff
@@ -26,6 +26,16 @@ type ContainerAttachConfig struct {
+// the chunks using this metadata.
+type LogMetaData struct {
+	IsPartial   bool
+	LastPartial bool
```
Removed IsPartial and renamed IsLastPartial to Last.

anusha-ragunathan commented on daemon/logger/copier.go:
```diff
@@ -47,6 +48,13 @@ func (c *Copier) Run() {
 
+func setLogMetaData(msg *Message, partialid string, ordinal int, lastPartial bool) {
+	msg.MetaData.IsPartial = true
+	msg.MetaData.ID = partialid
```
Yes, its the ID of a bunch of partials

cpuguy83 commented:
Keep in mind that generating an ID is not free.
I don't see why an ID would be needed for full messages unless there was some way to correlate that ID with something in the API.


Brian Goff

> On Apr 5, 2018, at 18:08, Sebastiaan van Stijn <notifications@github.com> wrote:
> 
> @thaJeztah commented on this pull request.
> 
> In daemon/logger/copier.go:
> 
> > @@ -112,7 +136,20 @@ func (c *Copier) copySrc(name string, src io.Reader) {
>  					msg.Source = name
>  					msg.Timestamp = time.Now().UTC()
>  					msg.Line = append(msg.Line, buf[p:n]...)
> -					msg.Partial = true
> +
> +					// Generate unique partialID for first partial. Use it across partials.
> +					// Initialize index for first partial. Increment it across partials.
> +					if firstPartial {
> +						partialid = stringid.GenerateRandomID()
> +						ordinal = 0
> +						firstPartial = false
> +					}
> +					ordinal++
> +					setLogMetaData(msg, partialid, ordinal, false)
> Let me try explaining my train of thought: if each message has a unique ID (irregardless of being partial or "full"), an aggregator could find all parts of a message by "grouping" by ID, and stitch them together based on ordinal.
> 
> The same logic could be used both for partial and full messages, so processing could be easier to handle if there's less differences between those
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

cpuguy83 reviewed (commented):
Wondering if the message timestamp for partials should all be the same?

anusha-ragunathan commented:
@cpuguy83 : Theoretically, they should be. But it doesnt serve any real purpose.

thaJeztah commented:
> Keep in mind that generating an ID is not free.

Fair point. Still a bit on the fence for "partials" starting with ordinal 1, and "full" messages having ordinal 0 (https://github.com/moby/moby/pull/35831#discussion_r177184442)

> Wondering if the message timestamp for partials should all be the same?

We're already using some state variables for partial messages; perhaps adding one for timestamp would work to address that? (even if it's theoretical, it would be "correct")

anusha-ragunathan commented:
@thaJeztah : I'll change the struct to a pointer to struct. That way, its nil for Full entries and avoids any confusion.

anusha-ragunathan commented:
@cpuguy83 : Updated PR to have same timestamp for all partials
@thaJeztah : Updated PR to have PartialLogMetaData only for partials. Full log entries have a nil pointer.

PTAL.

cpuguy83 commented on daemon/logger/copier.go:
```diff
@@ -94,9 +101,23 @@ func (c *Copier) copySrc(name string, src io.Reader) {
 					msg.Line = append(msg.Line, buf[p:p+q]...)
 
+					if hasMorePartial {
+						msg.PLogMetaData = &types.PartialLogMetaData{ID: partialid, Ordinal: ordinal, Last: true}
```
Can we keep a `PartialLogMetaData` around to re-use... or usage a sync/pool?
I think just keeping a single one should work.

anusha-ragunathan replied:
Since the copy is sequential, I'm now simply using one allocated object over all msgs. Would that work for you?

anusha-ragunathan replied:
Done!

cpuguy83 reviewed (approved):
LGTM

thaJeztah commented:
These are known flaky's, tracked through https://github.com/moby/moby/issues/36501

```
21:45:58 ----------------------------------------------------------------------
21:45:58 FAIL: docker_api_swarm_service_test.go:33: DockerSwarmSuite.TestAPIServiceUpdatePort
21:45:58 
21:45:58 [ddf763c4c66ae] waiting for daemon to start
21:45:58 [ddf763c4c66ae] daemon started
21:45:58 
21:45:58 docker_api_swarm_service_test.go:39:
21:45:58     waitAndAssert(c, defaultReconciliationTimeout, d.CheckActiveContainerCount, checker.Equals, 1)
21:45:58 docker_utils_test.go:452:
21:45:58     c.Assert(v, checker, args...)
21:45:58 ... obtained int = 0
21:45:58 ... expected int = 1
21:45:58 
21:45:58 [ddf763c4c66ae] exiting daemon
21:46:10 

22:10:50 ----------------------------------------------------------------------
22:10:50 FAIL: docker_cli_swarm_test.go:1605: DockerSwarmSuite.TestSwarmPublishDuplicatePorts
22:10:50 
22:10:50 [de0f01c91cd86] waiting for daemon to start
22:10:50 [de0f01c91cd86] daemon started
22:10:50 
22:10:50 docker_cli_swarm_test.go:1613:
22:10:50     // make sure task has been deployed.
22:10:50     waitAndAssert(c, defaultReconciliationTimeout, d.CheckActiveContainerCount, checker.Equals, 1)
22:10:50 docker_utils_test.go:452:
22:10:50     c.Assert(v, checker, args...)
22:10:50 ... obtained int = 0
22:10:50 ... expected int = 1
22:10:50 
22:10:50 [de0f01c91cd86] exiting daemon
22:10:53 
```


This one looks to have become flaky again (tracking issue:  https://github.com/moby/moby/issues/32673). I attached the daemon logs for that test
[docker.log.txt](https://github.com/moby/moby/files/1905216/docker.log.txt)


```
21:48:21 ----------------------------------------------------------------------
21:48:21 FAIL: docker_api_swarm_test.go:297: DockerSwarmSuite.TestAPISwarmLeaderElection
21:48:21 
21:48:21 [db7b38f7b479e] waiting for daemon to start
21:48:21 [db7b38f7b479e] daemon started
21:48:21 
21:48:21 [d0ba13e9ba563] waiting for daemon to start
21:48:21 [d0ba13e9ba563] daemon started
21:48:21 
21:48:21 [df5215dd8a143] waiting for daemon to start
21:48:21 [df5215dd8a143] daemon started
21:48:21 
21:48:21 [db7b38f7b479e] exiting daemon
21:48:21 assertion failed: error is not nil: Error response from daemon: rpc error: code = DeadlineExceeded desc = context deadline exceeded
21:48:21 [d0ba13e9ba563] exiting daemon
21:48:21 [df5215dd8a143] exiting daemon
```

thaJeztah reviewed (approved):
LGTM, thanks!

sharonx commented:
splunk-logging-plugin is already using the Partial flag now to reassemble messages. This change is a breaking change for our plugins. I assume this change would go in a major release?

cpuguy83 commented:
@sharonx this is an internal only change and opens up built-in drivers to support the additional parameters.

Plugins will require some extra work to expose this data to the plugin, but there should be no change to the data currently being sent to the plugin. When the work is done to bring plugins up to speed with thos change it will be purely additive.

anusha-ragunathan commented:
@kyaparla : Messages get split at 16K. So what you are observing is not expected.
Also, the issue you describe is not related to this PR. Please open a new issue and cc @sharonx and @chenziliang from Splunk.
